keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,389 @@
+extensions may not be as mature yet, or may be slow to transition to “ready” for legitimate reasons.
+
+The lack of a single unambiguous signal on when a “node” is “ready”, beyond just the kubelet or a container runtime
+is causing several issues today.
+
+* System components like Cluster Autoscaler and Scheduler have to combine several Node Conditions to evaluate a node’s
+actual readiness. It is not straightforward to introduce new conditions into the policy.
+* External systems (such as networking by the cloud provider) must often be configured before pods should be scheduled on a node.
+* Cluster Autoscaler incorrectly assumes a GPU node to be “ready” prior to device plugins and/or device drivers
+being installed which leads to spurious node additions and deletions thereby leading to custom logic in the autoscaler.
+* Logs & metrics are lost if logging & monitoring agents are not online which can lead to sub-optimal (often unexpected)
+user behavior.
+* A node with offline kube-proxy can result in service outage in the event of an application update (via deployments for example).
+* It is difficult for Node Lifecycle Controller and/or node repair systems to detect and remedy issues with extended system
+components of the node since they lack a set of generic & easy to identify signals from those individual components on the node.
+
+As Kubernetes becomes more and more customizable (or extensible) and gets deployed in mission critical systems,
+it is imperative to provide fail safe mechanisms that are built into the system.
+
+The remainder of this proposal introduces an approach for addressing the issue of node health extensibility using
+node readiness gates.
+
+### Goals
+
+* introduce a mechanism that allows for customizable/extensible node readiness checks.
+
+### Non-Goals
+
+* improving time-to-readiness of nodes
+
+## Proposal
+
+Any solution that is chosen for addressing the set of problems stated above should meet the following design requirements:
+
+* Be backwards compatible
+* Be extensible to accommodate variations in kubernetes environments - the set of system microservices can vary across k8s deployments.
+* Ideally, provide a single signal (not a policy) to cluster level components.
+* Ideally, avoid introducing new API primitives.
+* Support self-bootstrapping - scheduling system components (including the kubelet) even if the node is not logically ready.
+
+## Design Details
+
+### Determining Readiness via Pod Selectors
+
+In this design, the indication of node readiness can be expressed by the readiness/health of specific pods running on that node. This approach
+would leverage existing mechnanisms and well-known concepts to determine if a node is ready to receive pods.
+
+A readiness taint will be used to track the overall readiness of a node. The kubelet, node lifecycle controller, other external controllers will
+manage that taint.
+
+The Node object will be extended to include a list of readiness gates like so:
```

Comment:
> And in fact, it seems like really we already have a thing that selects nodes that need special initialization: DaemonSets. You could just add the BlocksReadiness field to DaemonSetSpec, defaulting to false. And then a node would be "meta-ready" when all readiness-blocking DaemonSets that selected that node had successfully scheduled their pods to that node and the pods reported being ready.

I like this idea but I do think there should still be a way to set the readiness gate outside of DaemonSets also. Maybe for an initial implementation we can only allow readiness gates from cloud providers and from DaemonSets and if other use-cases come up we can add support later?

Reply:
Should something like `blocksReadiness` on a DaemonSet be allowed only on namespaces like `kube-system`? With this, any user with access to DaemonSets could prevent pods from scheduling on the entire cluster.

Reply:
Ah... yeah, there would need to be some sort of protection there, but it can't be as simple as "limited to `kube-system`". I'm not sure what approach would be consistent with other features. ("Limited to namespaces with the annotation `X`?")